torch/csrc/distributed/c10d/TCPStoreLibUvBackend.cpp

```diff
@@ -16,12 +18,1210 @@
+        MAX_KEY_COUNT);
+
+    StreamWriter sw(iptr());
+    for (const auto _ : c10::irange(key_count)) {
+      (void)_; // Suppress unused variable warning
+      std::string key;
+      if (!stream.read_key(key)) {
+        return false;
+      }
+
+      auto data = store->get(key);
+      sw.write_vector(data);
+    }
+    sw.send();
+
+    return true;
+  }
+
+  bool parse_multi_set_command() {
+    uint64_t key_count = 0;
+    if (!stream.read_value(key_count)) {
+      return false;
+    }
+    TORCH_CHECK(
+        key_count <= MAX_KEY_COUNT,
+        "Too many keys with multi_get. keys:{} max:{}",
+        key_count,
+        MAX_KEY_COUNT);
+
+    for (const auto _ : c10::irange(key_count)) {
+      (void)_; // Suppress unused variable warning
+
+      std::string key;
+      if (!stream.read_key(key)) {
+        return false;
+      }
+
+      std::vector<uint8_t> newData;
+      if (!stream.read_payload(newData))
+        return false;
+      store->set(key, newData);
+    }
+
+    return true;
+  }
+
+  bool parse_cancel_wait_command() {
+    store->clearClientWaitState(iptr());
+
+    StreamWriter sw(iptr());
+    sw.write1((uint8_t)WaitResponseType::WAIT_CANCELED);
+    sw.send();
+
+    return true;
+  }
+
+ public:
+  explicit UvClient(uv_loop_t* loop, LibUVStoreDaemon* store)
+      : UvTcpSocket(loop), store(store) {}
+
+  static c10::intrusive_ptr<UvClient> make(
+      uv_loop_t* loop,
+      LibUVStoreDaemon* store) {
+    auto res = c10::make_intrusive<UvClient>(loop, store);
+    res->handleReady();
+    return res;
+  }
+
+  c10::intrusive_ptr<UvClient> iptr() {
+    return c10::intrusive_ptr<UvClient>::reclaim_copy(this);
+  }
+
+ protected:
+  void onClose() override {
+    store->unregisterClient(iptr());
+  }
+};
+
+void LibUVStoreDaemon::onConnect(int status) {
+  auto client = UvClient::make(&loop, this);
+  registerClient(client);
+  try {
+    tcpServer->accept(client);
+    client->startRead();
+  } catch (std::exception& e) {
+    C10D_INFO("Failed to accept client due to {}", e.what());
+    client->close();
+  }
+}
+
+void LibUVStoreDaemon::onExitRequest() {
+  C10D_DEBUG("Store exit requested\n");
+  uv_close((uv_handle_t*)&exit_handle, nullptr);
+  uv_stop(&loop);
+}
+
+void LibUVStoreDaemon::init(const TCPStoreOptions& opts) {
+  if (opts.masterListenFd.has_value()) {
+    tcpServer = UvTcpServer::makeWithSocket(&loop, *opts.masterListenFd);
+  } else {
+    try {
+      tcpServer = UvTcpServer::makeWithPort(&loop, opts.port, /*useIpv6=*/true);
+    } catch (std::exception& ex) {
+      C10D_INFO(
+          "Failed to bind to ipv6 address, trying ipv4. Error: {}", ex.what());
+      tcpServer =
+          UvTcpServer::makeWithPort(&loop, opts.port, /*useIpv6=*/false);
+    }
+  }
+  tcpServer->setOnConnectCallback(
+      [this](auto status) { this->onConnect(status); });
+
+  port_ = tcpServer->port();
+  TORCH_CHECK(
+      port_ == opts.port || opts.port == 0, // zero means use any port
```

Comment:
@H-Huang @XilunWu  this was the only change from the previous PR.
it did not check for `port == 0`.